Making the card component look good

Sbonelo01 opened the pull request:
Related issues: https://umuzi-projects.monday.com/boards/1274815998/pulses/1363777379

## Description:

Making the card component look good

## Note:
I am aware that a user profile picture can be uploaded in the backend. But for now, I am using static pictures for the avatars.

## Screenshots/videos

Before:

![image](https://github.com/Umuzi-org/Tilde/assets/37469171/d7a89a69-41b4-4fc1-98a1-62de06675a24)

After:

![tilde-card-7](https://github.com/Umuzi-org/Tilde/assets/37469171/f71dd138-5583-47b5-a6db-f729574be6d2)

## I solemnly swear that:

- [x] My code follows the style guidelines of this project
- [x] I have merged the develop branch into my branch and fixed any merge conflicts
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested new or existing tests and made sure that they pass

Sbonelo01 committed:
structure of the card sorted

Sbonelo01 committed:
static fields added

Sbonelo01 committed:
Poppins imported, tags and flavours traversed

Sbonelo01 committed:
remove comments

Sbonelo01 committed:
avatars

Sbonelo01 committed:
dates and status

Sbonelo01 committed:
cleanup

Sbonelo01 committed:
remove comments

Sbonelo01 committed:
pr displays correctly

Sbonelo01 committed:
touch ups

Sbonelo01 committed:
remove comments

Ngoakor12 reviewed (changes requested):
Implementation-wise, it looks very similar to the designs :ok_hand: 

- The card content doesn't seem to be centered. There is no breathing room on the right. The card details lean to the right leading to some of the card details being hidden. Is there no way to add breathing room and avoid hidden content? E.g. long project names?
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/eb48b913-65d3-4815-9dfc-8e93d843a44c)

- The content button is not centered. I'm guessing that's because you wanted to add an icon there, but if there isn't one available, I think the text should be centered.
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/4874de92-57f9-4c34-b02b-e06fa5f3655e)

- The whole button should be clickable, not just the text

- The due date shows in a funny way
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/38adf241-d655-4171-85de-d5dd212ddde6)
- The tooltips are not there anymore. Maybe they should be in some places? E.g. three dot icon, avatars


General questions:
- What does the date at the top represent? It's not obvious and seems confusing maybe it shouldn't be there?
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/d23dd3c8-283b-4381-a5f3-17f3d1a0fba0)

- What does this icon represent? What's meant to happen when it's clicked?
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/d3687e4b-39f6-4418-9613-840f0dd80def)

Sbonelo01 committed:
touch ups

Sbonelo01 committed:
adding tooltip

Sbonelo01 committed:
cleanup

Sbonelo01 committed:
reviewers sliced

Sbonelo01 committed:
remove options icon

Sbonelo01 committed:
adding comment for tag icon not being found on font awesome

Ngoakor12 commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,170 +1,223 @@
 {% load humanize %}
 
+<!-- TODO: Extract constants and store them theme.py -->
```
should this be here?

Ngoakor12 reviewed (changes requested):
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/eb55eb1f-3d18-4c3e-a7cc-5e33ca5b2f18)
- The review count icons are bunched up together, I'm not sure if that's intended or not.
- Also, the competent review count does not have an icon for some reason. 
- And these icons need tooltips, otherwise it's hard to make sense of them.

![image](https://github.com/Umuzi-org/Tilde/assets/54069197/e8a0306b-6661-4760-9120-5c18b5779935)
- This PR section needs a tooltip too


Side note:
- Regarding the tags. I understand you felt they were extra busy with the icon and description inside a chip, but I'd argue we still have the same thing with the hashtags in place of the icons. Personally I think the Tilde 2 original(tag inside a chip without an icon or hashtag) looks better and is less busy.

Sbonelo01 committed:
remove hashtags

Sbonelo01 committed:
tooltips

Sbonelo01 committed:
remove comment

Sbonelo01 commented:
@Ngoakor12 I've addressed your feedback
The tooltip for a bunch of reviewers doesn't make that much sense right now. I have an idea of what needs to be done. (Here is an issue on MoM: https://umuzi-projects.monday.com/boards/1274815998/pulses/1387241235)

Ngoakor12 commented:
> * The card content doesn't seem to be centered. There is no breathing room on the right. The card details lean to the right leading to some of the card details being hidden. Is there no way to add breathing room and avoid hidden content? E.g. long project names?
>   ![image](https://private-user-images.githubusercontent.com/54069197/300800813-eb48b913-65d3-4815-9dfc-8e93d843a44c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMDc4NDEsIm5iZiI6MTcwNzIwNzU0MSwicGF0aCI6Ii81NDA2OTE5Ny8zMDA4MDA4MTMtZWI0OGI5MTMtNjVkMy00ODE1LTlkZmMtOGU5M2Q4NDNhNDRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDA4MTkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0YzI5ZTYwNGFiZjZkMDgyY2JkOGFiYTk1ZjhlZGFmNGRlOWYwMjI4ODlhYTgzZmM0ZThjNWZjMWUzZTBlZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AcIwj4nhbVA0fOgetCKWfVf0sUy42HiKML614Pu59wY)

This point :point_up:  seems to have been undone or not fully resolved.

Screenshot from the last review:
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/3cba2194-f679-42d2-85fb-10521d7c4d81)

Screenshot now:
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/82abb7e6-9745-4fa1-916e-651805ed566a)

Ngoakor12 commented:
Nevermind, it seems to be a responsiveness/out of date chrome on linux browser issue

sheenarbw commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,170 +1,225 @@
+             {% endif %}
+             {% if card.code_review_excellent_since_last_review_request %}
+                <div class="relative inline-flex items-center p-2 text-sm font-medium text-center bg-opacity-10 rounded-lg">
+                  <i class="fa-regular fa-face-laugh-wink" aria-hidden="true" title="Excellent reviews"></i>
```
This is looking very repetitive. Can you try make it more DRY?

This could help: https://github.com/carltongibson/django-template-partials

sheenarbw replied:
If it seems like a difficult thing to do or it will cause too many delays, please add a new Mom task for it. You can label it as technical debt. We really should try hard not to write repetitive code.

Sbonelo01 replied:
Technical debt issue added: https://umuzi-projects.monday.com/boards/1274815998/pulses/1398535402

sheenarbw commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,170 +1,225 @@
+         </div>
+
+        {% if card.open_pr_count %}
+         <a href={{ card.repo_url }} title="Pull request info">
```
There was a request to be able to navigate straight to the repo from the card even if there are no open PRs. This request was made with Tilde1, might as well implement it now.

Ngoakor12 replied:
@Sbonelo01 The current workaround for this issue only works when there are open PRs. The previous card had a "Repository" text to achieve this. I don't know what needs to happen because the card design you are following does not consider this.

I think a GitHub icon that links to the repo if it exist should be fine for now, maybe there's a better way to do it.

sheenarbw commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,170 +1,225 @@
+          {% for assignee in card.assignee_names %}
+            <img
+              class="w-[30px] h-[30px] rounded-full"
+              src="https://images.unsplash.com/photo-1594751543129-6701ad444259?q=80&w=1000&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8Mnx8ZGFyayUyMHByb2ZpbGV8ZW58MHx8MHx8fDA%3D"
```
for now can you please make this an anonomous image/icon. We are unlikely to implement profile pics before we go live.

sheenarbw commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,170 +1,225 @@
+      
+      <div class="grid grid-cols-2 grid-auto-rows-1fr gap-x-10 gap-y-10 w-full h-[auto]">
+        <div class="Start-button col-span-1 row-span-1 w-full">
+          <button
```
Clicking on the content button should not start the card. It should rather link to the card content and open it in a new tab.   

Alternatively,  `alert("TODO")` so it's very obvious that it needs to be fixed, and you can also make a Mom task so it can be done

Please don't make buttons do random things!

sheenarbw commented on backend/frontend/theme.py:
```diff
@@ -1,10 +1,12 @@
     "button_primary_small": f"rounded bg-{PRIMARY_COLOUR}-600 px-2 py-1 text-xs font-semibold text-white shadow-sm hover:bg-{PRIMARY_COLOUR}-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-{PRIMARY_COLOUR}-600",
-    #
-    #
+    "button_content_small": f"rounded bg-{CONTENT_BTN_COLOUR} px-2 py-1 text-xs font-semibold text-white shadow-sm hover:bg-{CONTENT_BTN_COLOUR} focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-{CONTENT_BTN_COLOUR}",
```
Is this only used in one place? If yes, it does not belong in this file.

This file is about creating things we will use over and over again.  We can have many small primary buttons on many pages all over the place.  But content buttons are very rare.

sheenarbw replied:
Why not just make use of primary and secondary buttons? You can update the definitions. We don't want a new button style defined for every function that a button might do. We should reuse things.

sheenarbw commented on backend/frontend/theme.py:
```diff
@@ -1,10 +1,12 @@
 PRIMARY_COLOUR = "indigo"
+CONTENT_BTN_COLOUR = "indigo-950"
```
Please aim to be consistent with the things that are already in place. If you are defining a new standard it's good to be explicit about it. This just seems a bit random. This one colour constant will behave differently from all the rest.

I'm also not convinced it's necessary.

sheenarbw reviewed (changes requested):
This is looking good but there are a few rough edges.

Also, please always run the unit tests, even if you are doing pure frontend work. This is a big piece of code and there is a chance it'll break something.

Sbonelo01 committed:
Merge branch 'develop' into user_board/make_card_component_look_good

Sbonelo01 committed:
cleanup

Sbonelo01 committed:
Merge branch 'user_board/make_card_component_look_good' of github.com:Umuzi-org/Tilde into user_board/make_card_component_look_good

Sbonelo01 committed:
button conditions

Sbonelo01 committed:
button fixes

Sbonelo01 committed:
revert theme file

Sbonelo01 committed:
cleanup

Sbonelo01 committed:
comments

Ngoakor12 commented on backend/frontend/templates/frontend/user/board/partial_card.html:
```diff
@@ -1,180 +1,201 @@
+      </div>
+      {% if card.open_pr_count %}
+        {% with modified_url=card.repo_url|slice:"25:-4" %}
+          <a href="https://github.com/umuzi-org/{{ modified_url }}/pulls"
```
I think it's best if we reuse the `get_github_repo_link` method that was used before and just append "/pulls" at the end and avoid hardcoding the URL here.

Sbonelo01 committed:
reused get_github_repo_link method

Sbonelo01 committed:
github icon
